File: RayRacerKit/TIA.swift
//
//  TIA.swift
//  RayRacerKit
//
//  Created by Serge Tsyba on 29.6.2023.
//

public class TIA {
	private(set) public var players = (Player(), Player())
	private(set) public var missiles = (Missile(), Missile())
	private(set) public var ball = Ball()
	private(set) public var playfield = Playfield()
	
	private var colors = Array(repeating: 0, count: 4)
	private var collisions = 0
	
	public var peripheral: Peripheral = .none
	public var output: GraphicsOutput?
	
	private var input = 0x0
	private var buffer = Array(repeating: 0, count: 4*20)
	
	/// Indicates whether vertical sync is enabled.
	/// TIA outputs vertical sync signal when enabled, which is done by writing VSYNC register.
	private(set) public var isVerticalSyncEnabled: Bool = false {
		didSet {
			if self.isVerticalSyncEnabled {
				self.output?
					.sync(.vertical)
			}
		}
	}
	
	/// Indicates whether wait on color sync is currently inabled.
	/// TIA is currently waiting on horizontal sync.
	private(set) public var awaitsHorizontalSync: Bool = false
	
	/// Indicates whether vertical blanking is enabled.
	/// TIA outputs blank video signal when enabled, which is done by writing to VBLANK register.
	private(set) public var isVerticalBlankEnabled: Bool = false
	
	/// Indicates whether horizontal blanking is enabled.
	/// TIA outputs blank video signal when enabled, which occurrs automatically for the first 76 or 68
	/// color clocks of each scan line, depending on whether HMOVE register was strobed or not.
	public var isHorizontalBlankEnabled: Bool {
		return self.colorClock < self.horizontalBlankResetColorClock
	}
	
	/// Color clock at which horizontal blank is turned off.
	///
	/// Horizontal blank is reset at
	/// - color clock 68 normally
	/// - color clock 76 when HMOVE register strobed
	private var horizontalBlankResetColorClock = 68
	
	/// Color clock within the current scan line.
	private(set) public var colorClock: Int = 0 {
		didSet { self.colorClock %= 228 }
	}
	
	/// Resets TIA.
	public func reset() {
		self.isVerticalSyncEnabled = false
		self.isVerticalBlankEnabled = false
		self.awaitsHorizontalSync = false
		
		self.horizontalBlankResetColorClock = 68
		self.colorClock = 0
	}
	
	/// Advances color clock by 1 unit.
	public func advanceClock() {
		if self.colorClock == 0 {
			self.awaitsHorizontalSync = false
			self.horizontalBlankResetColorClock = 68
			
			self.output?
				.sync(.horizontal)
		}
		
		if self.isHorizontalBlankEnabled {
			// position counters of movable objects do not receive clock
			// signals during horizontal blank
			self.output?.blank()
		} else {
			let options = self.drawOptions(at: self.colorClock - 68)
			
			if self.isVerticalBlankEnabled {
				self.output?
					.blank()
			} else {
				let object = Self.objectIndexes[options.rawValue]
				let color = self.colors[object]
				self.output?
					.write(color: color)
			}
			
			// advance movable object position counters
			self.players.0.position += 1
			self.players.1.position += 1
			self.missiles.0.position += 1
			self.missiles.1.position += 1
			self.ball.position += 1
		}
		
		self.colorClock += 1
	}
}

extension TIA {
	public enum GraphicsSync {
		case vertical
		case horizontal
	}
	
	public protocol GraphicsOutput {
		/// Signals the start of a new field or scan line.
		mutating func sync(_ sync: GraphicsSync)
		/// Signals the absence of color for the next value.
		mutating func blank()
		/// Signals the next color value.
		mutating func write(color: Int)
	}
	
	public protocol Peripheral {
		func read() -> Int
	}
	
	/// One of six objects TIA draws on screen.
	protocol MovableObject {
		/// Position counter of this object.
		/// This object will need to be drawn at specific values of its position counter.
		var position: Int { get set }
		/// Amount of horizontal motion, which adjusts this object's position.
		/// Value ranges in [0, 15] and will move this object left or right on the current scan line once
		/// HMOVE register is strobed.
		var motion: Int { get set }
		/// Returns `true` when this object should be drawn; returns `false` otherwise.
		var needsDrawing: Bool { get }
	}
}

extension TIA {
	private static let objectIndexes: [Int] = (0x00...0xff)
		.map() {
			var options = DrawOptions(rawValue: $0)
			
			if options.contains(.player0) ||
				options.contains(.missile0) {
				// player 0/missile 0
				return 0
			} else if options.contains(.player1) ||
						options.contains(.missile1) {
				// player 1/missile 1
				return 1
			} else if options.contains(.ball) {
				// ball
				return 2
			} else if options.contains(.playfield) {
				// playefield
				if options.contains(.scoreMode) {
					// score mode: player 0 or player 1
					return options.contains(.player0) ? 0 : 1
				} else {
					// playfield
					return 2
				}
			} else {
				// background
				return 3
			}
		}
	
	private func drawOptions(at point: Int) -> DrawOptions {
		var options: DrawOptions = []
		if self.players.0.needsDrawing {
			options.insert(.player0)
		}
		if self.players.1.needsDrawing {
			options.insert(.player1)
		}
		if self.missiles.0.needsDrawing {
			//			options.insert(.missile0)
		}
		if self.missiles.1.needsDrawing {
			//			options.insert(.missile1)
		}
		if self.ball.needsDrawing {
			options.insert(.ball)
		}
		if self.playfield.draws(at: point) {
			options.insert(.playfield)
		}
		if self.playfield.control.contains(.scoreMode) {
			options.insert(.scoreMode)
		}
		if point < 80 {
			options.insert(.leftHalf)
		}
		return options
	}
	
	private static let collisionsLookUp: [Int] = (0x00...0x1f)
		.map() {
			var collisions = 0
			// cxm0p
			collisions[0] = $0[1] && $0[0]
			collisions[1] = $0[1] && $0[2]
			// cxm1p
			collisions[2] = $0[3] && $0[2]
			collisions[3] = $0[3] && $0[0]
			// cxp0fb
			collisions[4] = $0[0] && $0[4]
			collisions[5] = $0[0] && $0[5]
			// cxp1fb
			collisions[6] = $0[2] && $0[4]
			collisions[7] = $0[2] && $0[5]
			// cxm0fb
			collisions[8] = $0[1] && $0[4]
			collisions[9] = $0[1] && $0[5]
			// cxm1fb
			collisions[10] = $0[3] && $0[4]
			collisions[11] = $0[3] && $0[5]
			// cxblpf
			collisions[12] = $0[4] && $0[5]
			// cxppmm
			collisions[13] = $0[1] && $0[3]
			collisions[14] = $0[0] && $0[2]
			
			return collisions
		}
}

extension TIA {
	func applyHorizontalMotion() {
		// NOTE: in hardware, horizontal motion advances position counters
		// of movable objects gradually every 4 color clocks, beginning
		// approximately 7 color clocks after HMOVE register is strobed;
		// so, at most it takes 15*4+7=67 color clocks to apply maximum
		// horizontal motion value of 15;
		// when HMOVE is strobed late during horizontal blank interval or
		// during visible portion of a scan line, horizontal motion
		// results are unpredictable and depend on clock timings and
		// hardware model;
		// the following is a simplified emulation of horizontal motion:
		// - when HMOVE register is strobed early during horizontal blanking
		//   interval, as intended, it applies motion at once to reduce
		//   unnecessary calculations, since there is no graphics output
		//   anyway;
		// - when HMOVE register is strobed late during horizontal blanking
		//   interval, it applies as much motion as would fit into
		//   horizontal blanking interval, as if it was applied gradually,
		//   and ignores the remaining amount of horizontal motion;
		// - when HMOVE register is strobed during visible portion of a
		//   scan line, it completely ignores horizontal motion, since the
		//   vast majority of games never do that anyway;
		
		let remainingClock = (68+8)-7 - self.colorClock
		guard remainingClock > 0 else {
			// ignore horizontal motion when HMOVE strobed late during
			// horizontal blanking interval or during visible portion of
			// a scan line
			return
		}
		
		// calculate maximum amount of horizontal motion, which could be
		// applied during horizontal blanking interval
		let ripples = remainingClock / 4
		self.players.0.move(limit: ripples)
		self.players.1.move(limit: ripples)
		self.missiles.0.move(limit: ripples)
		self.missiles.1.move(limit: ripples)
		self.ball.move(limit: ripples)
	}
}


// MARK: -
// MARK: Bus integration
extension TIA: Addressable {
	public func read(at address: Int) -> Int {
		switch address % 0x10 {
		case 0x00:
			// MARK: CXM0P
			return ((self.collisions & 0x3) << 6) | address
		case 0x01:
			// MARK: CXM1P
			return ((self.collisions & 0xc) << 4) | address
		case 0x02:
			// MARK: CXP0FB
			return ((self.collisions & 0x30) << 2) | address
		case 0x03:
			// MARK: CXP1FB
			return (self.collisions & 0xc0) | address
		case 0x04:
			// MARK: CXM0FB
			return ((self.collisions & 0x300) >> 2) | address
		case 0x05:
			// MARK: CXM1FB
			return ((self.collisions & 0xc00) >> 4) | address
		case 0x06:
			// MARK: CXBLPF
			return ((self.collisions & 0x1000) >> 5) | address
		case 0x07:
			// MARK: CXPPMM
			return ((self.collisions & 0x6000) >> 7) | address
			
		case 0x08:
			// MARK: INPT0
			return (self.input << 7) & 0x80
		case 0x09:
			// MARK: INPT1
			return (self.input << 6) & 0x80
		case 0x0a:
			// MARK: INPT2
			return (self.input << 5) & 0x80
		case 0x0b:
			// MARK: INPT3
			return (self.input << 4) & 0x80
		case 0x0c:
			// MARK: INPT4
			let data = self.peripheral.read()
			return (data << 3) & 0x80
		case 0x0d:
			// MARK: INPT5
			let data = self.peripheral.read()
			return (data << 2) & 0x80
		default:
			return .random(in: 0x00..<0x100)
		}
	}
	
	public func write(_ data: Int, at address: Int) {
		switch address {
		case 0x00:	// MARK: VSYNC
			self.isVerticalSyncEnabled = data[1]
		case 0x01:	// MARK: VBLANK
			self.isVerticalBlankEnabled = data[1]
		case 0x02:	// MARK: WSYNC
			// NOTE: when last CPU clock cycle of a write instruction coincides
			// with last three TIA color clocks in a scan line, WSYNC will
			// incorrectly stay on for an extra scanline, since it is reset at
			// the end of each TIA color clock cycle emulation, but the writing
			// CPU clock cycle is executed after that in console clock
			// emulation;
			// ensuring color clock is not reset guards against this edge case
			self.awaitsHorizontalSync = self.colorClock > 0
		case 0x03:	// MARK: RSYNC
			self.colorClock = 0
			
		case 0x04:	// MARK: NUSIZ0
			self.players.0.copies = data & 0x7
			self.missiles.0.copies = data & 0x7
			self.missiles.0.size = 1 << ((data >> 4) & 0x3)
			
		case 0x05:	// MARK: NUSIZ1
			self.players.1.copies = data & 0x7
			self.missiles.1.copies = data & 0x7
			self.missiles.1.size = 1 << ((data >> 4) & 0x3)
			
		case 0x06:	// MARK: COLUP0
			self.colors[0] = data
		case 0x07:	// MARK: COLUP1
			self.colors[1] = data
		case 0x08:	// MARK: COLUPF
			self.colors[2] = data
		case 0x09:	// MARK: COLUBK
			self.colors[3] = data
			
		case 0x0a:	// MARK: CTRLPF
			self.playfield.control = Playfield.Control(rawValue: data & 0x3)
			self.ball.size = 1 << ((data >> 4) & 0x3)
		case 0x0d:	// MARK: PF0
			self.playfield.graphics[0] = UInt8(data)
		case 0x0e:	// MARK: PF1
			self.playfield.graphics[1] = UInt8(Int(reversingBits: data))
		case 0x0f:	// MARK: PF2
			self.playfield.graphics[2] = UInt8(data)
		case 0x0b:	// MARK: REFP0
			self.players.0.reflected = data[3]
		case 0x0c:	// MARK: REFP1
			self.players.1.reflected = data[3]
			
		case 0x10:	// MARK: RESP0
			self.players.0.reset()
		case 0x11:	// MARK: RESP1
			self.players.1.reset()
		case 0x12:	// MARK: RESM0
			self.missiles.0.reset()
		case 0x13:	// MARK: RESM1
			self.missiles.1.reset()
		case 0x14:	// MARK: RESBL
			self.ball.reset()
			
		case 0x1b:	// MARK: GRP0
			self.players.0.graphics.0 = UInt8(data)
			self.players.1.graphics.1 = self.players.1.graphics.0
		case 0x1c:	// MARK: GRP1
			self.players.1.graphics.0 = UInt8(data)
			self.players.0.graphics.1 = self.players.0.graphics.0
			self.ball.enabled.1 = self.ball.enabled.0
			
		case 0x1d:	// MARK: ENAM0
			self.missiles.0.enabled = data[1]
		case 0x1e:	// MARK: ENAM1
			self.missiles.1.enabled = data[1]
		case 0x1f:	// MARK: ENABL
			self.ball.enabled.0 = data[1]
			
		case 0x20:	// MARK: HMP0
			self.players.0.motion = (data >> 4) ^ 0x8
		case 0x21:	// MARK: HMP1
			self.players.1.motion = (data >> 4) ^ 0x8
		case 0x22:	// MARK: HMM0
			self.missiles.0.motion = (data >> 4) ^ 0x8
		case 0x23:	// MARK: HMM1
			self.missiles.1.motion = (data >> 4) ^ 0x8
		case 0x24:	// MARK: HMBL
			self.ball.motion = (data >> 4) ^ 0x8
			
		case 0x25:	// MARK: VDELP0
			self.players.0.delayed = data[0]
		case 0x26:	// MARK: VDELP1
			self.players.1.delayed = data[0]
		case 0x27:	// MARK: VDELBL
			self.ball.delayed = data[0]
			
		case 0x2a:	// MARK: HMOVE
			self.horizontalBlankResetColorClock = 68+8
			self.applyHorizontalMotion()
			
		case 0x2b:	// MARK: HMCLR
			self.players.0.motion = 0
			self.players.1.motion = 0
			self.missiles.0.motion = 0
			self.missiles.1.motion = 0
			self.ball.motion = 0
			
		case 0x2c:	// MARK: CXCLR
			self.collisions = 0
		default:
			break
		}
	}
}


// MARK: -
// MARK: Convenience functionality
public extension Int {
	init(reversingBits value: Int) {
		self = 0
		for bit in 0..<8 {
			self[bit] = value[7-bit]
		}
	}
}

private extension TIA.MovableObject {
	/// Resets position counter of this object.
	mutating func reset() {
		self.position = 160-4
	}
	
	/// Applies horizontal motion.
	/// This advances position counter of this object by the value of horizontal motion, but limited by
	/// the specified motion limit.
	mutating func move(limit: Int) {
		self.position += min(self.motion, limit)
	}
}


private struct DrawOptions: OptionSet {
	static let player0   = DrawOptions(rawValue: 1 << 0)
	static let player1   = DrawOptions(rawValue: 1 << 1)
	static let missile0  = DrawOptions(rawValue: 1 << 2)
	static let missile1  = DrawOptions(rawValue: 1 << 3)
	static let ball	     = DrawOptions(rawValue: 1 << 4)
	static let playfield = DrawOptions(rawValue: 1 << 5)
	static let scoreMode = DrawOptions(rawValue: 1 << 6)
	static let leftHalf  = DrawOptions(rawValue: 1 << 7)
	
	var rawValue: Int
	
	init(rawValue: Int) {
		self.rawValue = rawValue
	}
}


// MARK: -
private extension TIA.Peripheral where Self == NoPeripheral {
	static var none: Self {
		return NoPeripheral()
	}
}

private struct NoPeripheral: TIA.Peripheral {
	func read() -> Int {
		return .random(in: 0x00...0xff)
	}
}

File: librayracer/tia.c
//
//  tia.c
//  rayracer
//
//  Created by Serge Tsyba on 19.11.2025.
//
#include "tia.h"
#include "flags.h"

#include <stdlib.h>

// MARK: Object positioning
static void advance_player_position(racer_player *player) {
	player->position += 1;

	if (player->position == 160) {
		player->position = 0;
		// reset position counter of a missile, if it is reset to player
		*player->missile_position = 0;

		// clear position reset flag and enable drawing main copy
		player->control &= ~PLAYER_POSITION_RESET;
		player->copy_mask |= 0x1;
	}
}

#define advance_object_position(object) \
	(object)->position += 1; \
	if ((object)->position == 160) { \
		(object)->position = 0; \
	}

#define apply_object_motion(object) \
	(object)->position += (object)->motion ^ 0x8

#define reset_object_motion(object) \
	(object)->motion = 0

static uint16_t copy_modes[][2] = {
	{0x001, 0},	// ●○○○○○○○○○
	{0x005, 0},	// ●○●○○○○○○○
	{0x011, 0},	// ●○○●○○○○○○
	{0x015, 0},	// ●○●○●○○○○○
	{0x101, 0},	// ●○○○○○○○●○
	{0x001, 1},	// ●●○○○○○○○○
	{0x111, 0},	// ●○○○●○○○●○
	{0x001, 2}	// ●●●●○○○○○○
};


// MARK: -
void racer_tia_reset(racer_tia *tia) {
	tia->color_clock = 0;
	*tia->is_ready = true;
	tia->blank_reset_clock = 68;

	tia->output_control = 0x00;
	tia->input_control = 0x00;
	tia->input_latch = 0xc0;

	tia->players[0].missile_position = &null_missile_position;
	tia->players[1].missile_position = &null_missile_position;

	// TODO: send composite sync
}

void racer_tia_advance_clock(racer_tia *tia) {
	// NOTE: scan line reset check needs to happen at the beginning of
	// a color clock cycle due to simultaneous clock simulation of
	// the console
	if (tia->color_clock >= 228) {
		tia->color_clock = 0;
		*tia->is_ready = true;
		tia->blank_reset_clock = 68;

		// notify video output horizontal sync started
		tia->video_sync(tia->video_output, VIDEO_HORIZONTAL_SYNC);
	}

	const bool horizontal_blank = tia->color_clock < tia->blank_reset_clock;
	uint8_t color = horizontal_blank;

	// position counters of movable objects do not receive clock signals
	// during horizontal blanking/retrace; no need to re-calculate draw state
	// and update object collisions
	if (!horizontal_blank) {
		const uint16_t state = get_object_draw_state(tia);

		// set output color unless TIA ouputs blank
		const bool vertical_blank = tia->output_control & TIA_OUTPUT_VERTICAL_BLANK;
		color |= vertical_blank;

		if (!vertical_blank) {
			const int index = draw_indices[state];
			color = tia->colors[index];
		}

		// update collisions
		tia->collisions |= collisions[state >> 3];

		// advance position counters of graphics objects
		advance_player_position(&tia->players[0]);
		advance_player_position(&tia->players[1]);
		advance_object_position(&tia->missiles[0]);
		advance_object_position(&tia->missiles[1]);
		advance_object_position(&tia->ball);
	}

	tia->color_clock += 1;

	// sync video output when buffer is filled
	if (tia->video_buffer == tia->video_buffer_end) {
		tia->video_sync(tia->video_output, VIDEO_BUFFER_SYNC);
	}

	// write color output
	*tia->video_buffer = color;
	tia->video_buffer++;
}


// MARK: -
// MARK: Input port
void racer_tia_write_port(racer_tia *tia, uint8_t data) {
	// latch 0 on pins 4,5 when port latch enabled
	if (!(tia->input_control & TIA_INPUT_PORT_LATCH)) {
		tia->input_latch &= (data & 0xc0);
	}
}


// MARK: -
// MARK: Bus
uint8_t racer_tia_read(const racer_tia *tia, uint8_t address) {
	switch (address % 0x10) {
		case 0x00: {// MARK: cxm0p
			const uint8_t data = (tia->collisions >> 0) & 0x3;
			return (data << 6) | address;
		}
		case 0x01: {// MARK: cxm1p
			const uint8_t data = (tia->collisions >> 2) & 0x3;
			return (data << 6) | address;
		}
		case 0x02: {// MARK: cxp0fb
			const uint8_t data = (tia->collisions >> 4) & 0x3;
			return (data << 6) | address;
		}
		case 0x03: {// MARK: cxp1fb
			const uint8_t data = (tia->collisions >> 6) & 0x3;
			return (data << 6) | address;
		}
		case 0x04: {// MARK: cxm0fb
			const uint8_t data = (tia->collisions >> 8) & 0x3;
			return (data << 6) | address;
		}
		case 0x05: {// MARK: cxm1fb
			const uint8_t data = (tia->collisions >> 10) & 0x3;
			return (data << 6) | address;
		}
		case 0x06: {// MARK: cxblpf
			const uint8_t data = (tia->collisions >> 12) & 0x1;
			return (data << 6) | address;
		}
		case 0x07: {// MARK: cxppmm
			const uint8_t data = (tia->collisions >> 14) & 0x3;
			return (data << 6) | address;
		}

		case 0x08: {// MARK: inpt0
			const uint8_t data = tia->read_port(tia->peripheral);
			return (data << 7) & 0x80;
		}
		case 0x09: {// MARK: inpt1
			const uint8_t data = tia->read_port(tia->peripheral);
			return (data << 6) & 0x80;
		}
		case 0x0a: {// MARK: inpt2
			const uint8_t data = tia->read_port(tia->peripheral);
			return (data << 5) & 0x80;
		}
		case 0x0b: {// MARK: inpt3
			const uint8_t data = tia->read_port(tia->peripheral);
			return (data << 4) & 0x80;
		}
		case 0x0c: {// MARK: inpt4
			const uint8_t data = (tia->input_control & TIA_INPUT_PORT_LATCH)
			? tia->input_latch
			: tia->read_port(tia->peripheral);

			return (data << 3) & 0x80;
		}
		case 0x0d: {// MARK: inpt5
			const uint8_t data = (tia->input_control & TIA_INPUT_PORT_LATCH)
			? tia->input_latch
			: tia->read_port(tia->peripheral);

			return (data << 2) & 0x80;
		}
		default:
			return arc4random_uniform(0x100);
	}
}

void racer_tia_write(racer_tia *tia, uint8_t address, uint8_t data) {
	switch (address) {
		case 0x00: {// MARK: vsync
			const bool vertical_sync = data & 0x2;
			set_flag(tia->output_control, VIDEO_VERTICAL_SYNC, vertical_sync);

			// notify video output when vertical sync started
			if (vertical_sync) {
				tia->video_sync(tia->video_output, VIDEO_VERTICAL_SYNC);
			}
			break;
		}

		case 0x01: {// MARK: vblank
			// vertical blanking
			const bool vertical_blank = data & 0x2;
			set_flag(tia->output_control, TIA_OUTPUT_VERTICAL_BLANK, vertical_blank);

			// input control
			tia->input_control = data & 0xc0;
			// reset input latches when input port I4-I5 latching is disabled;
			// both values are reset to 1
			if (!(tia->input_control & TIA_INPUT_PORT_LATCH)) {
				tia->input_latch = 0x30;
			}
			break;
		}

		case 0x02:	// MARK: wsync
			// when the last clock cycle of WSYNC write instruction coincides
			// with the last color clock of a scan line (which resets color
			// clock to 0), WSYNC should not be enabled
			if (tia->color_clock != 0) {
				*tia->is_ready = false;
			}
			break;

		case 0x03:	// MARK: rsync
			// FIXME: RSYNC
			tia->color_clock = -6;
			break;

		case 0x04: {// MARK: nusiz0
			const uint16_t *copy_mode = copy_modes[data & 0x7];
			tia->players[0].copy_mask = copy_mode[0];
			tia->players[0].scale = copy_mode[1];

			const int missile_scale = (data >> 4) & 0x3;
			tia->missiles[0].copy_mask = copy_mode[0];
			tia->missiles[0].size = 1 << missile_scale;
			break;
		}
		case 0x05: {// MARK: nusiz1
			const uint16_t *copy_mode = copy_modes[data & 0x7];
			tia->players[1].copy_mask = copy_mode[0];
			tia->players[1].scale = copy_mode[1];

			const int missile_scale = (data >> 4) & 0x3;
			tia->missiles[1].copy_mask = copy_mode[0];
			tia->missiles[1].size = 1 << missile_scale;
			break;
		}

		case 0x06:	// MARK: colup0
			tia->colors[0] = data;
			break;
		case 0x07:	// MARK: colup1
			tia->colors[1] = data;
			break;
		case 0x08:	// MARK: colupf
			tia->colors[2] = data;
			break;
		case 0x09:	// MARK: colubk
			tia->colors[3] = data;
			break;

		case 0x0a: {// MARK: ctrlpf
			tia->playfield.control = data & 0x3;
			tia->ball.size = 1 << ((data >> 4) & 0x3);
			break;
		}

		case 0x0d: {// MARK: pf0
			const uint64_t graphics = data >> 4;
			tia->playfield.graphics[0] &= 0xffff0ffff0;
			tia->playfield.graphics[0] |= graphics | (graphics << 20);

			const uint64_t reflected = reflections[graphics];
			tia->playfield.graphics[1] &= 0x0fffffff0;
			tia->playfield.graphics[1] |= graphics | (reflected << (40-8));
			break;
		}
		case 0x0e: {// MARK: pf1
			const uint64_t graphics = reflections[data];
			tia->playfield.graphics[0] &= 0xff00fff00f;
			tia->playfield.graphics[0] |= (graphics << 4) | (graphics << (20+4));

			const uint64_t reflected = data;
			tia->playfield.graphics[1] &= 0xf00ffff00f;
			tia->playfield.graphics[1] |= (graphics << 4) | (reflected << (20+8));
			break;
		}
		case 0x0f: {// MARK: pf2
			const uint64_t graphics = data;
			tia->playfield.graphics[0] &= 0x00fff00fff;
			tia->playfield.graphics[0] |= (graphics << 12) | (graphics << (20+12));

			const uint64_t reflected = reflections[data];
			tia->playfield.graphics[1] &= 0xfff0000fff;
			tia->playfield.graphics[1] |= (graphics << 12) | (reflected << 20);
			break;
		}

		case 0x0b:	// MARK: refp0
			set_flag(tia->players[0].control, PLAYER_REFLECTED, !(data & 0x8));
			break;
		case 0x0c:	// MARK: refp1
			set_flag(tia->players[1].control, PLAYER_REFLECTED, !(data & 0x8));
			break;

		case 0x10: {// MARK: resp0
			// it takes 4 color clock cycles to reset position counter and
			// an extra clock cycle to latch the draw start signal
			tia->players[0].position = 160-4-1;

			// clear first bit in copy mask to skip drawing first player copy
			tia->players[0].control |= PLAYER_POSITION_RESET;
			tia->players[0].copy_mask &= ~0x1;
			break;
		}
		case 0x11: {// MARK: resp1
			tia->players[1].position = 160-4-1;
			tia->players[1].control |= PLAYER_POSITION_RESET;
			tia->players[1].copy_mask &= ~0x1;
			break;
		}
		case 0x12:	// MARK: resm0
			// it takes 4 color clock cycles to reset position counter
			tia->missiles[0].position = 160-4;
			break;
		case 0x13:	// MARK: resm1
			tia->missiles[1].position = 160-4;
			break;
		case 0x14:	// MARK: resbl
			tia->ball.position = 160-4;
			break;

		case 0x1b: {// MARK: grp0
			// set player 0 graphics
			tia->players[0].graphics[0] = data;
			tia->players[0].graphics[1] = reflections[data];

			// copy player 1 delayed graphics
			tia->players[1].graphics[2] = tia->players[1].graphics[0];
			tia->players[1].graphics[3] = tia->players[1].graphics[1];
			break;
		}

		case 0x1c: {// MARK: grp1
			// set player 1 graphics
			tia->players[1].graphics[0] = data;
			tia->players[1].graphics[1] = reflections[data];

			// copy player 0 delayed graphics
			tia->players[0].graphics[2] = tia->players[0].graphics[0];
			tia->players[0].graphics[3] = tia->players[0].graphics[1];

			// copy ball delayed control flag
			tia->ball.control &= ~BALL_ENABLED_1;
			tia->ball.control |= (bool)(tia->ball.control & BALL_ENABLED_0);
			break;
		}

		case 0x1d:	// MARK: enam0
			set_flag(tia->missiles[0].control, MISSILE_ENABLED, data & 0x2);
			break;
		case 0x1e:	// MARK: enam1
			set_flag(tia->missiles[1].control, MISSILE_ENABLED, data & 0x2);
			break;
		case 0x1f:	// MARK: enabl
			set_flag(tia->ball.control, BALL_ENABLED_0, data & 0x2);
			break;

		case 0x20:	// MARK: hmp0
			tia->players[0].motion = data >> 4;
			break;
		case 0x21:	// MARK: hmp1
			tia->players[1].motion = data >> 4;
			break;
		case 0x22:	// MARK: hmm0
			tia->missiles[0].motion = data >> 4;
			break;
		case 0x23:	// MARK: hmm1
			tia->missiles[1].motion = data >> 4;
			break;
		case 0x24:	// MARK: hmbl
			tia->ball.motion = data >> 4;
			break;

		case 0x25:	// MARK: vdelp0
			set_flag(tia->players[0].control, PLAYER_DELAYED, data & 0x1);
			break;
		case 0x26:	// MARK: vdelp1
			set_flag(tia->players[1].control, PLAYER_DELAYED, data & 0x1);
			break;
		case 0x27:	// MARK: vdelbl
			set_flag(tia->ball.control, BALL_DELAYED, data & 0x1);
			break;

		case 0x28: {// MARK: resmp0
			if (data & 0x2) {
				tia->missiles[0].control |= MISSILE_RESET_TO_PLAYER;
				tia->players[0].missile_position = &tia->missiles[1].position;
			} else {
				tia->missiles[0].control &= ~MISSILE_RESET_TO_PLAYER;
				tia->players[0].missile_position = &null_missile_position;
			}
			break;
		}
		case 0x29: {// MARK: resmp1
			if (data & 0x2) {
				tia->missiles[1].control |= MISSILE_RESET_TO_PLAYER;
				tia->players[1].missile_position = &tia->missiles[1].position;
			} else {
				tia->missiles[1].control &= ~MISSILE_RESET_TO_PLAYER;
				tia->players[1].missile_position = &null_missile_position;
			}
			break;
		}

		case 0x2a: {// MARK: hmove
			tia->blank_reset_clock = 68+8;

			// NOTE: in hardware, horizontal motion is applied gradually,
			// 1 unit every 4 color clocks, approximately 7 color clocks
			// after HMOVE register is strobed;
			// this is a simplified simulation which applies horizontal
			// motion at once, it will produce correct results when HMOVE
			// is strobed at the beginning of a scan line, as recommended
			// by Atari 2600 programming manual
			apply_object_motion(&tia->players[0]);
			apply_object_motion(&tia->players[1]);
			apply_object_motion(&tia->missiles[0]);
			apply_object_motion(&tia->missiles[1]);
			apply_object_motion(&tia->ball);
			break;
		}

		case 0x2b: {// MARK: hmclr
			reset_object_motion(&tia->players[0]);
			reset_object_motion(&tia->players[1]);
			reset_object_motion(&tia->missiles[0]);
			reset_object_motion(&tia->missiles[1]);
			reset_object_motion(&tia->ball);
			break;
		}

		case 0x2c:	// MARK: cxclr
			tia->collisions = 0;
			break;

		default:
			break;
	}
}
